Attempt to support garbage collected arrays/vectors

Techcable opened the pull request:
This crashes even the most basic tests :(
Everything is broken.

I feel like I need to clean up the allocator interface even more.

NOTE: This supersedes #20

Techcable committed:
First attempt a garbage collected arrays/vectors

This crashes even the most basic tests :(
Everything is broken.

I feel like I need to cleanup.

andersk commented:
Although this is marked as superseding #20, it won’t fix #15 (support garbage collected trait objects).

Techcable committed:
Fix GcArray SEGFAULTs with the standard allocator

Using the small-object-arenas allocator still causes
a SEGFAULT unfortunately.
However, I've made significant progress ;)

Techcable committed:
Fix incorrect layouts in small-object arena alloc

As of now, the 'array' tests pass.

Add cfg!(zerogc_simple_debug_alloc) which adds internal padding bytes.
This was very helpful in finding this bug ;)

Techcable committed:
Respect explicit where bounds in unsafe_gc_impl!

They should suppress the generation of implicit where bounds

Techcable committed:
Remove GcRebrand + GcErase for GcVec

They are unsafe, because GcVec has an explicit reference
to the owning context, which cant be meaningfully erased.

Add GcVec::extend_from_slice

Techcable committed:
[simple] Write a test for GcVec

Mostly a copy of the "arrays" test.
It's even in the same file.

Techcable committed:
Actually allocate an empty vec instead of dangling

Dangling pointers don't have a valid `GcVecHeader`,
so its not possible to fetch the length/capacity.

The allocated vector is a singleton (assuming proper alignment),
so we still wont waste much memory.

Techcable committed:
When growing a GcVec, actually copy the memory

Also remove the Drop impl from GcRawVec.
Dropping is the responsibility of the underlying RawVecRepr.

Techcable committed:
Correct the layout in alloc_vec_with_capacity

We were including the header in the 'value_layout',
which should only be the array value....

Techcable committed:
Trace the singleton empty vec

If we want to keep using it, we need the garbage
collector to preserve it....

Fix GcVec::extend_from_slice to correctly do set_len

Techcable commented:
You are right @andersk . I'll put that next on my TODO list :)
